src/asl.ts

```diff
@@ -688,15 +692,40 @@ export class ASL {
               ],
               Default: "exit",
             },
-            assign: {
-              Type: "Pass",
-              node: stmt.expr,
-              // in a forIn, the variable is the index, we will rewrite `var` to `$.{var}.index`
-              // `arr[var]` is rewritten to `$.{var}.item`
-              InputPath: `${tempArrayPath}[0]`,
-              ResultPath: `$.${stmt.variableDecl.name}`,
-              Next: "body",
-            },
+            /**
+             * Assign the index to $.[variableName].
+             * When the loop.variableDecl is an {@link Identifier} (not {@link VariableStmt}), the variable may be used after the for loop.
+             */
+            assign: isForOfStmt(stmt)
+              ? {
+                  Type: "Pass",
+                  node: stmt.variableDecl,
+                  InputPath: `${tempArrayPath}[0]`,
+                  ResultPath: `$.${stmt.variableDecl.name}`,
+                  Next: "body",
+                }
+              : /**ForInStmt
+                 * Assign the value to $.0__[variableName].
+                 * Assign the index to the variable decl. If the variable decl is an identifier, it may be carried beyond the ForIn.
+                 */
+                {
+                  startState: "assignIndex",
+                  node: stmt.variableDecl,
+                  states: {
+                    assignIndex: {
+                      Type: "Pass",
+                      InputPath: `${tempArrayPath}[0].index`,
+                      ResultPath: `$.${stmt.variableDecl.name}`,
+                      Next: "assignValue",
+                    },
+                    assignValue: {
+                      Type: "Pass",
+                      InputPath: `${tempArrayPath}[0].item`,
+                      ResultPath: `$.0__${stmt.variableDecl.name}`,
+                      Next: "body",
+                    },
+                  },
```

Comment:
So, I had it as one step, it assigned `$.[var]` to the `{index, item}` object and then all reference were re-written as `$.[var].item` or `$.[var].index`.

It is not possible to set two top level variables in one state in ASL. you can overwrite the entire top level (`ResultPath: "$"`) or set a single property in the top level (`ResultPath: "$.[var]"`).

This didn't work when the `var` was an identifier with a lifetime beyond the for loop though. 

```ts
let i = "";
i = `${i}something`;
for(i in [1,2,3]) {
   ...
}
return i;
```

Because now the for loop would overwrite the `i` variable with an object instead of the index string.

One option would have been to "fix" the variable at the end of for loop (`$.[var] = $.[var].index`), but I decided to keep the variable clean (`$.[var]` always is the index of the for loop iteration, or the value in the case of a `ForOfStmt`.)

Options:
1. Current State - `$.[var]` (index) and `$.0__[var]` (item) - set in 2 states on each iteration.
2. Object and Fix - `$.[var] = { index, item }` for each iteration, after the for loop `$.[var] = $.[var].index`. - Variable may be shadowed internally, want to keep the model simple, prefer not to "fix" after.
3. Do not support using an identifier in `ForIn` - Prefer to support all TS features, no matter the cost.
4. Different behavior for `Identifier` and `VariableStmt` - Thought this would be too messy

Reply:
>Do not support using an identifier in ForIn - Prefer to support all TS features, no matter the cost

Is the cost only incurred in particular circumstances, or for all `for` loops?

Reply:
Only `for-in` from what I can see. Why does the value need to be set? Can't we replace all references to `array[i]` with the JSON path to the value?